src/opentimelineio/composition.h

```diff
@@ -72,12 +78,170 @@ class Composition : public Item {
 
+    size_t _bisect_right(
+        RationalTime const& tgt,
+        std::function<RationalTime(Composable*)> const& key_func,
+        ErrorStatus* error_status,
+        optional<size_t> lower_search_bound = optional<size_t>(0),
+        optional<size_t> upper_search_bound = nullopt) const;
+    size_t _bisect_left(
+        RationalTime const& tgt,
+        std::function<RationalTime(Composable*)> const& key_func,
+        ErrorStatus* error_status,
+        optional<size_t> lower_search_bound = optional<size_t>(0),
+        optional<size_t> upper_search_bound = nullopt) const;
+
     std::vector<Retainer<Composable>> _children;
     
     // This is for fast lookup only, and varies automatically
     // as _children is mutated.
     std::set<Composable*> _child_set;
 };
 
+template<typename T>
+inline std::vector<SerializableObject::Retainer<T>> Composition::each_child(
+    ErrorStatus* error_status,
+    optional<TimeRange> search_range,
+    bool shallow_search) const
+{
+    std::vector<Retainer<Composable>> children;
+    if (search_range)
+    {
+        auto range_map = range_of_all_children(error_status);
+        if (!error_status)
+            return {};
+
+        // find the first item whose end_time_inclusive is after the
+        // start_time of the search range
+        const auto first_inside_range = _bisect_left(
+            search_range->start_time(),
+            [&range_map](Composable* child) { return range_map[child].end_time_inclusive(); },
+            error_status);
+        if (!error_status)
+            return {};
+
+        // find the last item whose start_time is before the
+        // end_time_inclusive of the search_range
+        const auto last_in_range = _bisect_right(
+            search_range->end_time_inclusive(),
+            [&range_map](Composable* child) { return range_map[child].start_time(); },
+            error_status,
+            first_inside_range);
+        if (!error_status)
+            return {};
+
+        // limit the search to children who are in the search_range
+        for (auto child = _children.begin() + first_inside_range; child < _children.begin() + last_in_range; ++child)
+            children.push_back(child->value);
+    }
+    else
+    {
+        // otherwise search all the children
+        children = _children;
+    }
+    std::vector<Retainer<T>> out;
+    for (const auto& child : children)
+    {
+        if (auto valid_child = dynamic_cast<T*>(child.value))
+            out.push_back(valid_child);
+
+        // if not a shallow_search, for children that are compositions,
+        // recurse into their children
+        if (!shallow_search)
+            if (auto composition = dynamic_cast<Composition*>(child.value))
+            {
+                if (search_range)
+                {
+                    search_range = transformed_time_range(*search_range, composition, error_status);
```

Comment:
In some of these checks, the loop is aborted, and an empty return results if error_status is true. error_status reports things such as "no media reference set on clip."

1 - when is an error not an error? For example, if I have just created a clip, and not yet set a media reference, is that clip in an erroneous state? The existing implementation says yes, but the default construction is an an erroneous object. Is it sensible for OTIO to construct by default, erroneous things? This question's resolution is outside the scope of this PR, but it leads to

2- would I legitimately want to use these iterators to collect clips and test for being in an erroneous state, as part of a sanity checking tool?

3- even if there is an erroneous clip, would I still want all the non-erroneous clips?

4- should there be another argument to specify whether erroneous clips result in an empty return? In other words, might I want versions of this routine that are (a) all or nothing, (b) all, or (c) only non-erroneous? TBH, I'm only seeing the utility of (b). If an error_status resulted in the (a) case, I need to find some other way to iterate to discover the bad results, yet here we are in an otherwise lovely iterator...

5- is the above the behavior of the python equivalent, and ... same set of questions :)

Reply:
Good points! I should have looked closer at the error codes, I was assuming they were all fatal and should stop any further processing.

Taking a look at the Python code it looks like they throw exceptions (otio_errorStatusHandler.cpp)? The OTIO docs mention that exceptions shouldn't be used in C++ code so maybe that's another complication...

So the (b) option would be to not return immediately on an error, and continue processing? That does seem useful for uses cases like the sanity checking tool, though if there are multiple errors the user would only know about the last one.

Reply:
Throwing python exceptions is the responsibility of Python code :) Indeed the C++ code shouldn't throw, but that doesn't constrain the Python implementation.

I'd suggest, continue processing. If an error was encountered, return a new error status, perhaps "one or more invalid children encountered".